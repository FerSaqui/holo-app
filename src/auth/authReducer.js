/**
 * En un reducer, la acciÃ³n va a modificar el state y react va a renderiza/actualizar lo que tenga que actualizar
 */

import { types } from "../types/types";

export const authReducer = (state = {}, action) => {

    switch ( action.type ) {
        case types.login:
            return {
                ...action.payload,
                logged: true
            }

        case types.logout:
            return {
                logged: false
            }
    
        default:
            return state;
    }

}