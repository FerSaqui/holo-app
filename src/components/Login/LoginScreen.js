import React, { useContext } from 'react'
import { useNavigate } from 'react-router-dom';
import { AuthContext } from '../../auth/autContext';
import { authReducer } from '../../auth/authReducer';
import { types } from '../../types/types';

export const LoginScreen = () => {
  const navegacion = useNavigate();

  const {dispatch} = useContext(AuthContext);

  const handleLogin = () => {
    /**
     * Obtener del localstorage la ultima pagina antes de cerrar sesión.
     */

    const lastPage = localStorage.getItem("lastPage") || "/";

    const userAcction = {
      type: types.login,
      payload: {
        name: "Fernando",
      }
    };
    
    dispatch(userAcction);
    
    /**
     * Para redirigir a una pantalla especifica, donde:
     * Primer argumento: La ruta a redireccionar.
     * Segundo argumento: Un objeto que puede tener la propiedad replace en true para indicar que no pueda regresar
     *                    a la página anterior.
     */
    navegacion(lastPage, {
      replace: true //Reemplaza la vista actual o que reemplace la historia en lugar de crear una nueva entrada en la historia
    });
  }

  return (
    <div className="container my-5">
        <h1>Login Screen</h1>
        <hr />

        <button
          className="btn btn-primary"
          onClick={handleLogin}
        >
          Login
        </button>

    </div>
  )
}
