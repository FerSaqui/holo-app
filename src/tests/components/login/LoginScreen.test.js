import { mount} from "enzyme";
import { MemoryRouter, Route, Routes } from "react-router-dom";
import { AuthContext } from "../../../auth/autContext";
import { LoginScreen } from "../../../components/Login/LoginScreen";
import { types } from "../../../types/types";

const dispatch = jest.fn();

/**Simular el hook o la librería de react-router-dom */
const mockNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
    ...jest.requireActual("react-router-dom"),
    useNavigate: () => mockNavigate
}));

describe('Pruebas en <LoginComponent />', () => {
    const contextValue = {
        user: {
            logged: false
        },
        dispatch
    }

    const wrapper = mount(
        <AuthContext.Provider value={contextValue}>
            <MemoryRouter initialEntries={["/login"]}>
                <Routes>
                    <Route path="/login" element={<LoginScreen />} />
                </Routes>
            </MemoryRouter>
        </AuthContext.Provider>
    );

    test('Debe de mostrarse correctamente', () => {
        expect( wrapper ).toMatchSnapshot();
    });

    test('Debe de realizar el dispatch y la navegación', () => {
        const handleClick = wrapper.find("button").prop("onClick");
        handleClick();

        expect( contextValue.dispatch ).toHaveBeenCalledWith({
            type: types.login,
            payload: {name: "Fernando"}
        });

        expect( mockNavigate ).toHaveBeenCalledWith("/", {"replace": true});
        localStorage.setItem("lastPage", "/holoJP");
        
        handleClick();

        expect( mockNavigate ).toHaveBeenCalledWith("/holoJP", {"replace": true});
    });

});