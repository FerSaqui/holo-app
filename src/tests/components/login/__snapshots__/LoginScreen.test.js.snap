// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <LoginComponent /> Debe de mostrarse correctamente 1`] = `
<div
  className="container my-5"
>
  <h1>
    Login Screen
  </h1>
  <hr />
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Login
  </button>
</div>
`;
